primitives/bridge/src/container_token_to_ethereum_message_exporter.rs

```diff
@@ -0,0 +1,371 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+use {
+    alloc::vec::Vec,
+    core::{iter::Peekable, marker::PhantomData, slice::Iter},
+    frame_support::{ensure, traits::Get},
+    parity_scale_codec::{Decode, Encode},
+    snowbridge_core::{AgentId, ChannelId, TokenId, TokenIdOf},
+    snowbridge_outbound_queue_primitives::v1::message::{Command, Message, SendMessage},
+    sp_core::{H160, H256},
+    sp_runtime::traits::MaybeEquivalence,
+    xcm::latest::SendError::{MissingArgument, NotApplicable, Unroutable},
+    xcm::prelude::*,
+    xcm_executor::traits::{ConvertLocation, ExportXcm},
+};
+
+pub struct EthereumBlobExporter<
+    UniversalLocation,
+    EthereumNetwork,
+    OutboundQueue,
+    AgentHashedDescription,
+    ConvertAssetId,
+    BridgeChannelInfo,
+>(
+    PhantomData<(
+        UniversalLocation,
+        EthereumNetwork,
+        OutboundQueue,
+        AgentHashedDescription,
+        ConvertAssetId,
+        BridgeChannelInfo,
+    )>,
+);
+
+impl<
+        UniversalLocation,
+        EthereumNetwork,
+        OutboundQueue,
+        AgentHashedDescription,
+        ConvertAssetId,
+        BridgeChannelInfo,
+    > ExportXcm
+    for EthereumBlobExporter<
+        UniversalLocation,
+        EthereumNetwork,
+        OutboundQueue,
+        AgentHashedDescription,
+        ConvertAssetId,
+        BridgeChannelInfo,
+    >
+where
+    UniversalLocation: Get<InteriorLocation>,
+    EthereumNetwork: Get<NetworkId>,
+    OutboundQueue: SendMessage<Balance = u128>,
+    AgentHashedDescription: ConvertLocation<H256>,
+    ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+    BridgeChannelInfo: Get<Option<(ChannelId, AgentId)>>,
+{
+    type Ticket = (Vec<u8>, XcmHash);
+
+    fn validate(
+        network: NetworkId,
+        _channel: u32,
+        universal_source: &mut Option<InteriorLocation>,
+        destination: &mut Option<InteriorLocation>,
+        message: &mut Option<Xcm<()>>,
+    ) -> SendResult<Self::Ticket> {
+        let expected_network = EthereumNetwork::get();
+        let universal_location = UniversalLocation::get();
+
+        log::info!("validate params: network={network:?}, _channel={_channel:?}, universal_source={universal_source:?}, destination={destination:?}, message={message:?}, ");
+
+        if network != expected_network {
+            log::trace!(target: "xcm::ethereum_blob_exporter", "skipped due to unmatched bridge network {network:?}.");
+            return Err(NotApplicable);
+        }
+
+        // Cloning destination to avoid modifying the value so subsequent exporters can use it.
+        let dest = destination.clone().take().ok_or(MissingArgument)?;
+        if dest != Here {
```

Comment:
we should pass the para-id to the converter and dissallow anything that is sending a token that does not belong to that para